### PR TITLE
[AMDGPU] Document amdgpu-as in AMDGPUUsage

### PR DESCRIPTION
**Pull request by Pierre-vh**
Add a section about fence & address spaces that covers amdgpu-as.

### CODE REVIEW

File: llvm/docs/AMDGPUUsage.rst [hunk 1]
```diff
@@ -5969,6 +5969,31 @@ following sections:
...
+.. _amdgpu-fence-as:
+
+Fence and Address Spaces
+++++++++++++++++++++++++++++++
+
+LLVM fences do not have address space information, thus, fence
+codegen usually needs to be conservative and fence all address spaces.
+
+In the case of OpenCL, where synchronization can only happen in the
+same address space, this can result in extra unnecessary waits.
```
Comment (arsenm) [inline llvm/docs/AMDGPUUsage.rst:5981]: > can only happen in the same address space

Do you mean something more like "only needs to happen in specific address spaces"?

Comment (ssahasra) [reply llvm/docs/AMDGPUUsage.rst:5981]: Or to be more explicit, "only needs to happen in user-specified address spaces".

File: llvm/docs/AMDGPUUsage.rst [hunk 2]
```diff
@@ -5969,6 +5969,31 @@ following sections:
...
+For instance, a fence that is supposed to only target local memory will
```
Comment (ssahasra) [inline llvm/docs/AMDGPUUsage.rst:5982]: Use verb "synchronize" instead of "target".

File: llvm/docs/AMDGPUUsage.rst [hunk 3]
```diff
@@ -5969,6 +5969,31 @@ following sections:
...
+also have to wait on all global memory operations, which is unnecessary.
+
+:doc:`Memory Model Relaxation Annotations <MemoryModelRelaxationAnnotations>` can
+be used as an optimization hint for fences to solve this problem.
+The AMDGPU backend handles the following tags on fences:
```
Comment (ssahasra) [inline llvm/docs/AMDGPUUsage.rst:5987]: s/handles/recognizes/

File: llvm/docs/AMDGPUUsage.rst [hunk 4]
```diff
@@ -5969,6 +5969,31 @@ following sections:
...
+
+- ``amdgpu-as:local`` - fence only the local address space
+- ``amdgpu-as:global``- fence only the global address space
+
+This can avoid unnecessary waiting in many cases. However, those annotations are
+attached using metadata, which can always be dropped by the optimizer when it
+inhibits optimizations, and the cost of not performing that optimization is
```
Comment (ssahasra) [inline llvm/docs/AMDGPUUsage.rst:5994]: What is this paragraph trying to say to the reader? Is there something actionable for the reader to take care of?

**Review by ssahasra — CHANGES_REQUESTED** [review]

**Review by nhaehnle — COMMENTED** [review]
Thanks. I think this looks good.

As noted offline, I do think we should also accept `amdgpu-as` on atomic operations for consistency, but that's something for a different change.

File: llvm/docs/AMDGPUUsage.rst (hunk 4 above)
Comment (Pierre-vh) [reply llvm/docs/AMDGPUUsage.rst]: I'm trying to make it clear that the tags aren't a promise, just an optimization hint.
I changed the wording, is that better?

File: llvm/docs/AMDGPUUsage.rst [hunk 5]
```diff
@@ -5969,6 +5969,33 @@ following sections:
...
 
+.. _amdgpu-fence-as:
+
+Fence and Address Spaces
+++++++++++++++++++++++++++++++
+
+LLVM fences do not have address space information, thus, fence
+codegen usually needs to conservatively synchronize all address spaces.
+
+In the case of OpenCL, where fences only needs to synchronize
```
Comment (ssahasra) [inline llvm/docs/AMDGPUUsage.rst:5980]: "needs" should be "need".

**Review by ssahasra — APPROVED** [review]
LGTM. But please do a quick spelling/grammar check.

**Pull request merged**